k2/csrc/array_ops.h

```diff
@@ -380,6 +406,40 @@ Array2<T> ToContiguous(const Array2<T> &src);
 template <typename T>
 bool Equal(const Array2<T> &a, const Array2<T> &b);
 
+/*
+  Index `src` with `indexes`, as in src[indexes].
+     @param [in] src   Array whose elements are to be read
+     @param [in] indexes  Indexes into `src`; must satisfy
+                       `0 <= indexes[i] < indexes.Dim()` if
+                       `allow_minus_one == true`,
+                       else -1 is also allowed and the corresponding
+                       output element will be zero.
+     @return  Returns an `Array1<T>` of dimension indexes.Dim(),
+               with `ans[i] = src[indexes[i]]` (or zero if
+               `allow_minus_one == true` and `indexes[i] == 0`).
+ */
+template <typename T>
+Array1<T> Index(const Array1<T> &src, const Array1<int32_t> &indexes,
+                bool allow_minus_one);
+
+
+/*
+  Index `src` with `indexes`, as in src[indexes].
+     @param [in] src   Array whose elements are to be read
+     @param [in] indexes  Indexes into `src`; must satisfy
+                       `0 <= indexes[i] < indexes.Dim0()` if
+                       `allow_minus_one == true`,
+                       else -1 is also allowed and the corresponding
+                       output element will be zero.
+     @return  Returns an `Array2<T>` of shape (indexes.Dim(), src.Dim1()),
+               with `ans[i,j] = src[indexes[i,j]]` (or zero if
```

Comment:
`ans[i,j] = src[indexes[i], j]`? seems this function is to do a  row-index/select? maybe it's better to add documentation at the first line, or use another function name (e.g. IndexRow or something else, I'm not sure)?

Reply:
MM.. I just called it Index because it's the same thing as what [] does in Python.  But yes, IndexRow() makes sense. 
Can you change that and fix the other typos?
Regarding testing <0 instead of -1: it saves one instruction, it's an optimization because the only -ve number allowed is -1.

Reply:
IndexRows() would be fine as a name.

Reply:
OK, will do.